File: java/iterator/EmployeeSubOrdinateIterator.java
package iterator;

public class EmployeeSubOrdinateIterator implements  EmployeeHierarchyIterator {
    private final EmployeeHierarchyCollection employeeHierarchyCollection;
    private int currentSubOrdinatePosition = 0;

    public EmployeeSubOrdinateIterator(EmployeeHierarchyCollection employeeHierarchyCollection) {
        this.employeeHierarchyCollection = employeeHierarchyCollection;
    }

    @Override
    public boolean hasNext() {
        return currentSubOrdinatePosition < employeeHierarchyCollection.getEmployees().size();
    }

    @Override
    public Employee getNext() {
        System.out.println("Iterating through a graph DFS / BFS");
        if (!hasNext()) {
            return null;
        }
        return employeeHierarchyCollection.getEmployees().get(currentSubOrdinatePosition++);
    }
}

File: typescript/patterns/iterator/employee-subordinate-iterator.ts
class EmployeeSubOrdinateIterator implements EmployeeHierarchyIterator {
  private employeeHierarchyCollection: EmployeeHierarchyCollection;
  private currentSubOrdinatePosition: number = 0;

  constructor(employeeHierarchyCollection: EmployeeHierarchyCollection) {
    this.employeeHierarchyCollection = employeeHierarchyCollection;
  }

  getNext(): Employee | null {
    console.log("Iterating through a graph DFS / BFS");
    if (!this.hasNext()) return null;
    return this.employeeHierarchyCollection.getEmployees()[
      this.currentSubOrdinatePosition++
    ];
  }

  hasNext(): boolean {
    return (
      this.currentSubOrdinatePosition <
      this.employeeHierarchyCollection.getEmployees().length
    );
  }
}
